### PR TITLE
feat: Introduce migration log file.

### PR DESCRIPTION
**Pull request by SandPod**
### Changes
- Introduces a database migration registry that is serialized to disk.
  - The registry keeps track of what order the migrations should be applied.
  - The registry ensures that collisions in version control appear if multiple branches introduce a migration at the same time.

### Minor refactors:
- Introduce a database connection method that runs a simple query to determine if we are connected to the database. (Isolates the database connection check to a single failure).
- Moves the error message creation for creating migrations up into the command interface. (All UX determined the same place.)

### Guidance for review.
Grouped a couple of refactors up from that enables the introduction of the migration registry. The tests introduce right now are all e2e tests to ensure the top level functionality.

## Pre-launch Checklist

- [x] I read the [Contribute](https://docs.serverpod.dev/contribute) page and followed the process outlined there for submitting PRs.
- [x] This update contains only one single feature or bug fix and nothing else. (If you are submitting multiple fixes, please make multiple PRs.)
- [x] I read and followed the [Dart Style Guide](https://dart.dev/guides/language/effective-dart/style) and formatted the code with [dart format](https://dart.dev/tools/dart-format).
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`), and made sure that the documentation follows the same style as other Serverpod documentation. I checked spelling and grammar.
- [x] I added new tests to check the change I am making.
- [x] All existing and new tests are passing.
- [x] Any breaking changes are documented below.

If you need help, consider asking for advice on the [discussion board](https://github.com/serverpod/serverpod/discussions).

## Breaking changes

None - Modifies new functionality that is introduced in 1.2.

### CODE REVIEW

File: tests/serverpod_test_server/test_e2e_migrations/migration_registry_test.dart [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+        tag: tag,
+      );
+      expect(
+        createMigrationExitCode,
+        0,
+        reason: 'Failed to create migration, exit code was not 0.',
+      );
+      migrationRegistry = await MigrationTestUtils.loadMigrationRegistry();
+      expect(migrationRegistry, hasLength(migrationsBefore + 3));
+      expect(migrationRegistry.getLatest(), contains(tag));
```
Comment (Isakdl) [inline tests/serverpod_test_server/test_e2e_migrations/migration_registry_test.dart:129]: Suggestion: Validate all three entries.

**Review by Isakdl — CHANGES_REQUESTED** [review]
Looking good! 😄 

Some smaller comments!

File: packages/serverpod/lib/src/database/migrations/migrations.dart [hunk 2]
```diff
@@ -1,28 +1,29 @@
 import 'dart:io';
 
 import 'package:path/path.dart' as path;
+import 'package:serverpod/protocol.dart';
+import 'package:serverpod_shared/serverpod_shared.dart';
 
 /// Shared methods for migration versioning.
 class MigrationVersions {
   /// Provides a list of all available migration versions of a module.
+  static const _fileNameRegistryJson = 'migration_registry.json';
```
[the remark this answers is not in the document]
Comment (SandPod) [reply packages/serverpod/lib/src/database/migrations/migrations.dart:10]: Yes it is, this is part of the duplicate code. But will extract it and reference the same place.

File: packages/serverpod/lib/src/database/migrations/migrations.dart [hunk 3]
```diff
@@ -1,28 +1,29 @@
...
-        if (entity is Directory) {
-          versions.add(path.basename(entity.path));
-        }
-      }
-      versions.sort();
-      return versions;
-    } catch (e) {
+      return migrationRegistry?.migrations ?? [];
+    } on MigrationRegistryLoadException catch (_) {
+      // TODO: add error handling
```
[the remark this answers is not in the document]
Comment (SandPod) [reply packages/serverpod/lib/src/database/migrations/migrations.dart]: This as implemented in a later commit.

**Review by Isakdl — APPROVED** [review]
LGTM

**Pull request merged**